algoliasearch/src/test/java/com/algolia/search/saas/IndexTest.java

```diff
@@ -1113,4 +1116,57 @@ public void doRequestCompleted(JSONObject content, AlgoliaException error) {
             }
         });
     }
+
+    @Test
+    public void hostStatusDefaultIsUnknown() throws Exception {
+        List<String> hostsArray = (List<String>) Whitebox.getInternalState(client, "readHosts");
+        String randomHostName = getRandomString() + "-dsn.algolia.biz";
+        final String nextHostName = hostsArray.get(1);
+
+        // Given a first host that timeouts, randomized to ensure no system caching
+        hostsArray.set(0, randomHostName);
+        Whitebox.setInternalState(client, "readHosts", hostsArray);
+
+
+        // Expect reachable hosts before any connection
+        assertTrue("Hosts should be considered up before first connection.", client.isUpOrCouldBeRetried(randomHostName));
+        assertTrue("Hosts should be considered up before first connection.", client.isUpOrCouldBeRetried(nextHostName));
+
+        // Expect success after a failing host
+        testSearchAsync();
+
+        HashMap<String, AbstractClient.HostStatus> statuses = (HashMap<String, AbstractClient.HostStatus>) Whitebox.getInternalState(client, "hostStatuses");
+
+        // Expect down host after failed connection, up host after successful connection
+        assertFalse("A host that has failed recently should be considered down.", client.isUpOrCouldBeRetried(randomHostName));
+        assertTrue("A host that has succeeded recently should be considered up.", client.isUpOrCouldBeRetried(nextHostName));
+
+        hostsArray = (List<String>) Whitebox.getInternalState(client, "readHosts");
+        randomHostName = getRandomString() + "-dsn.algolia.biz";
+
+        // Given a short host delay and a first host that timeouts
+        final int delay = 100;
+        client.setHostDownDelay(delay);
+        hostsArray.set(0, randomHostName);
+        Whitebox.setInternalState(client, "readHosts", hostsArray);
+
+        // Expect success after a failing host
+        testSearchAsync();
+
+        // Expect
+        Thread.sleep(delay);
+        assertTrue("A host that has failed should be considered up once the delay is over.", client.isUpOrCouldBeRetried(randomHostName));
+    }
+
+
+    // Should not reconnect before delay to the same host twice if it fails
+    // Should reconnect after delay to the same host twice if it fails
+
+    private String getRandomString() {
+        String randomAppId;
+        byte[] array = new byte[7]; // length is bounded by 7
+        new Random().nextBytes(array);
+        randomAppId = new String(array, Charset.forName("UTF-8"));
```

Comment:
- I would be really surprised if this call succeeded in all cases. Not every sequence of bytes is valid UTF-8.
- Besides, using arbitrary characters in URLs is probably not a good idea, even if now many parts of Unicode are accepted (like Chinese Han characters, for example).

To solve both issues, I would prefer using a hexadecimal representation.